### PR TITLE
Don't nest child loggers

### PR DESCRIPTION
**Pull request by anthony-murphy**
Previously we were nesting child loggers which may have a performance impact, as a deeply nested child logger would have to call each nest logger up the stack creating new object along the way to build up the final event. This change flattens at each layer so child logger are never nested in each other.

### CODE REVIEW

File: packages/utils/telemetry-utils/src/logger.ts [hunk 1]
```diff
@@ -249,6 +249,32 @@ export class ChildLogger extends TelemetryLogger {
         namespace?: string,
         properties?: ITelemetryProperties,
         propertyGetters?: ITelemetryPropertyGetters): TelemetryLogger {
+        // if we are creating a child of a child, rather than nest, which will increase
+        // the callstack overhead, just generate a new logger that includes everything from the previous
+        if (baseLogger instanceof ChildLogger) {
+            const combinedProperties =
+                baseLogger.properties === undefined && properties === undefined ?
+                    undefined
+                    : {
```
Comment (heliocliu) [inline packages/utils/telemetry-utils/src/logger.ts:258]: nit: here and below put the ? and : in the same relevant spot so either
```
? undefined
: {
```
or
```
undefined :
{
```

File: packages/utils/telemetry-utils/src/test/childLogger.spec.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
+/*!
```
Comment (heliocliu) [inline packages/utils/telemetry-utils/src/test/childLogger.spec.ts:1]: yey tests

**Review by heliocliu — APPROVED** [review]

File: packages/utils/telemetry-utils/src/test/childLogger.spec.ts [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+            "test2",
+            {
+                testProperty2: true,
+            });
+
+        childLogger2.send({ category:"generic", eventName:"testEvent" });
+        assert(sent, "event should be sent");
+    });
+
+    it("Getters Are Combined",()=>{
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/test/childLogger.spec.ts:109]: Test case for overriding a property or getter?

File: packages/utils/telemetry-utils/src/logger.ts [hunk 4]
```diff
@@ -249,6 +249,32 @@ export class ChildLogger extends TelemetryLogger {
...
         propertyGetters?: ITelemetryPropertyGetters): TelemetryLogger {
+        // if we are creating a child of a child, rather than nest, which will increase
+        // the callstack overhead, just generate a new logger that includes everything from the previous
+        if (baseLogger instanceof ChildLogger) {
+            const combinedProperties =
+                baseLogger.properties === undefined && properties === undefined
+                    ? undefined
+                    : {
+                        ...baseLogger.properties,
+                        ... properties,
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:260]:
```suggestion
                        ...properties,
```

File: packages/utils/telemetry-utils/src/logger.ts [hunk 5]
```diff
@@ -249,6 +249,32 @@ export class ChildLogger extends TelemetryLogger {
...
+                    };
+            const combinedGetters =
+                baseLogger.propertyGetters === undefined && propertyGetters === undefined
+                    ? undefined
+                    : {
+                        ...baseLogger.propertyGetters,
+                        ... propertyGetters,
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:267]:
```suggestion
                        ...propertyGetters,
```

File: packages/utils/telemetry-utils/src/logger.ts [hunk 6]
```diff
@@ -249,6 +249,32 @@ export class ChildLogger extends TelemetryLogger {
...
+                    };
+
+            return new ChildLogger(
+                baseLogger.logger,
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:271]: Is there a good way to test that a grandchild logger's base logger is the grandparent? I see that baseLogger is protected, so there'd need to be some trickery.  Fine to punt, just checking.

**Review by markfields — APPROVED** [review]
🎂

So the point is to avoid mulitple calls to `TelemetryLogger.prepareEvent`, right?  Looks good.

File: packages/utils/telemetry-utils/src/logger.ts (hunk 4 above)
Comment (anthony-murphy) [reply packages/utils/telemetry-utils/src/logger.ts:260]: if we care about this there should be a lint rule

**Pull request merged**